### PR TITLE
Separate LivreData library from core and lib

### PR DESCRIPTION
**Pull request by tribal-tec**
Also some performance fixes and cleanups in the traverser.

### CODE REVIEW

File: livre/data/DFSTraversal.cpp [hunk 1]
```diff
@@ -17,53 +17,30 @@
...
-            _state.setVisitChild(true);
-            return false;
-        }
+        if (depth == 1)
+            return;
 
         const NodeIds& nodeIds = nodeId.getChildren();
         for (const NodeId& childNodeId : nodeIds)
-        {
             traverse(childNodeId, depth - 1, visitor);
```
Comment (eile) [inline livre/data/DFSTraversal.cpp]: Traversal on the higher nodes does not abort if a lower node returns false!?

Comment (tribal-tec) [inline livre/data/DFSTraversal.cpp]: it does with the visit() from line 34?!

Comment (eile) [inline livre/data/DFSTraversal.cpp]: It does it on the parent where a child requests abortion. The grand-parent and above will continue its sibling traversal, as far as I understand the code.

Comment (tribal-tec) [inline livre/data/DFSTraversal.cpp]: Don't see why you could stop traversing the rest of the siblings if one stops traversing. What do you want to have changed?

**Comment by tribal-tec** [comment]
What's missing/wrong to get an approval?

**Comment by rdumusc** [comment]
Didn't have time to review it yet, will do in a moment

**Comment by eile** [comment]
Not done with the review yet.

File: livre/data/LODNode.cpp [hunk 2]
```diff
@@ -54,13 +55,13 @@ void LODNode::_computeWorldBox(const Vector3ui& levelTotalBlockDimensions)
     lBoxCoordMin = lBoxCoordMin / levelTotalBlockDimensions[index];
     lBoxCoordMax = lBoxCoordMax / levelTotalBlockDimensions[index];
 
-    _worldBox = Boxf(lBoxCoordMin, lBoxCoordMax);
+    return Boxf(lBoxCoordMin, lBoxCoordMax);
 }
 
-void LODNode::_initialize()
+Boxui LODNode::_computeLocalBox()
```
Comment (eile) [inline livre/data/LODNode.cpp]: const?

File: livre/data/LODNode.cpp [hunk 3]
```diff
@@ -31,21 +31,22 @@ LODNode::LODNode(const NodeId& nodeId, const Vector3ui& blockSize,
                  const Vector3ui& levelTotalBlockDimensions)
     : _nodeId(nodeId)
     , _blockSize(blockSize)
+    , _localVoxelBox(_computeLocalBox())
+    , _worldBox(_computeWorldBox(levelTotalBlockDimensions))
 {
-    _initialize();
-    _computeWorldBox(levelTotalBlockDimensions);
+    _computeLocalBox();
```
Comment (eile) [inline livre/data/LODNode.cpp]: Has no side effects?

File: livre/data/LODNode.cpp (hunk 2 above)
Comment (tribal-tec) [inline livre/data/LODNode.cpp]: yes

File: livre/data/LODNode.cpp (hunk 3 above)
Comment (tribal-tec) [inline livre/data/LODNode.cpp]: that call is useless here

**Comment by tribal-tec** [comment]
that's not good :)

File: livre/data/LODNode.h [hunk 4]
```diff
@@ -104,21 +104,24 @@ class LODNode
...
-    void _initialize(); //!< Computes the internal values of LODBox (voxel
-                        //! extents, block extents, etc ).
-    void _computeWorldBox(
-        const Vector3ui& levelTotalBlockDimensions); //!< compute world box.
+    const NodeId _nodeId;       //!< Node id.
+    const Vector3ui _blockSize; //!< Number of voxels in each dimension without
+                                //! paddings.
+    const Boxui _localVoxelBox; //!< Voxelwise position and dimension in
+                                //! reflevel
+    //! resolution in model space.
```
Comment (rdumusc) [inline livre/data/LODNode.h]: mv to previous line?

File: livre/data/LODNode.h [hunk 5]
```diff
@@ -104,21 +104,24 @@ class LODNode
...
+    const Boxf _worldBox; //!< World space box in normalized coordinates [ 0,
+                          //! 1.0 ).
+    Boxui _computeLocalBox() const; //!< Computes the internal values of LODBox
+                                    //!(voxel
+                                    //! extents, block extents, etc ).
```
Comment (rdumusc) [inline livre/data/LODNode.h]: same

File: livre/data/NodeVisitor.h [hunk 6]
```diff
@@ -42,7 +42,7 @@ class NodeVisitor
      * @param renderNode is the node being visited.
      * @param state Visit state.
      */
-    virtual void visit(const NodeId& nodeId, VisitState& state) = 0;
+    virtual bool visit(const NodeId& nodeId) = 0;
```
Comment (rdumusc) [inline livre/data/NodeVisitor.h]: doc for return value

File: tests/lib/lodSelection.cpp [hunk 7]
```diff
@@ -19,12 +19,12 @@
 
 #define BOOST_TEST_MODULE LodSelection
 
-#include <livre/lib/data/MemoryDataSource.h>
+#include <livre/data/MemoryDataSource.h>
```
Comment (rdumusc) [inline tests/lib/lodSelection.cpp]: join with includes below

File: livre/data/LODNode.h (hunk 4 above)
Comment (tribal-tec) [inline livre/data/LODNode.h]: clang-format. good job. will try to fix

**Comment by tribal-tec** [comment]
Updated

**Pull request merged**